src/room_data.c

```diff
@@ -5199,8 +5199,11 @@ void destroy_dungeon_heart_room(PlayerNumber plyr_idx, const struct Thing *heart
     if (room_is_invalid(room) || (!room_role_matches(room->kind,RoRoF_KeeperStorage)))
     {
         WARNLOG("The heart thing is not in heart room");
-        long i = dungeonadd->room_kind[RoK_DUNGHEART];
-        room = room_get(i);
+        if (dungeonadd->backup_heart_idx == 0)
+        {
+            long i = dungeonadd->room_kind[find_first_roomkind_with_role(RoRoF_KeeperStorage)];
```

Comment:
find_first_roomkind_with_role is also not ideal if there would be multiple rooms with the role

Reply:
what function would you suggest? I checked them all and non of them seemed to be what I want.